### PR TITLE
[Reland] NVFP4 Marlin fallback for non-Blackwell GPUs

### PR DESCRIPTION
**Pull request by Godmook**
## Motivation

Reland of #19652.

The original PR had merge conflicts with recent main changes. This PR cleanly re-applies the same feature set with conflict resolution.

## Modifications

- **17 files changed** (same 16 as #19652 + 1 bug fix)
- All details and discussion available in the original PR.
- **Bug fix** (reported in #22459): `routed_scaling_factor` was applied twice in the MoE Marlin fallback path — once in TopK routing and again in `moe_sum_reduce`. Fixed by gating `should_fuse_routed_scaling_factor_in_topk` to `False` when Marlin fallback is active. This fix does not affect existing native FP4 (Blackwell) behavior.
- Changed fallback log messages from `warning` to `info` with a note about INT4 AWQ as an alternative. (Because as you can see benchmark results below, at dense model AWQ shows better performance)

## Accuracy Tests

Benchmarks run after the `routed_scaling_factor` bug fix:

**Llama-3.1-8B-Instruct (Dense) on A100-80GB** — lm_eval, 0-shot, 3 seeds avg

| Model | GSM8K | GPQA | ARC |
|-------|-------|------|-----|
| BF16 | 83.9% | 32.2% | 55.6% |
| INT4 AWQ | 80.3% | 32.6% | 53.8% |
| NVFP4 Marlin | 78.6% | 29.9% | 53.6% |

**Qwen3-30B-A3B (MoE) on A100-80GB** — SGLang eval, thinking mode

| Model | GPQA (repeat 8) | GSM8K |
|-------|----------------|-------|
| BF16 | 62.9% | 90.7% |
| NVFP4 Marlin | 62.2% | 91.1% |
| INT4 AWQ | 59.6% | 91.5% |

For dense models, INT4 AWQ is better. For MoE, NVFP4 Marlin is competitive with BF16 and outperforms AWQ on GPQA (+2.6%p). Whether to reland depends on how we weigh the dense model accuracy tradeoff against MoE benefits and the convenience of running official NVIDIA NVFP4 checkpoints out-of-the-box on non-Blackwell GPUs.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

---
### CI States

Latest PR Test (Base): :x: [Run #26702483338](https://github.com/sgl-project/sglang/actions/runs/26702483338)
Latest PR Test (Extra): :x: [Run #26702483322](https://github.com/sgl-project/sglang/actions/runs/26702483322)

### CODE REVIEW

**Comment by Godmook** [comment]
cc. @BBuf @Fridge003

**Comment by BBuf** [comment]
@b8zhong Is it ok now?

**Review by b8zhong — COMMENTED** [review]
Hi, apologies for the delay. My main concern is that it will introduce extra complexity into real NVFP4 code path, I'm OK with it if it can entirely contained in the Marlin files. Is it possible to move the Marlin FP4 for MoE entirely to be self contained (i.e, not touching modelopt_quant, etc)?

File: docs/references/environment_variables.md [hunk 1]
```diff
@@ -123,6 +123,7 @@ SGLang supports various environment variables that can be used to configure its
 | --- | --- | --- |
 | `SGLANG_INT4_WEIGHT` | Enable INT4 weight quantization | `false` |
 | `SGLANG_FORCE_FP8_MARLIN` | Force using FP8 MARLIN kernels even if other FP8 kernels are available | `false` |
+| `SGLANG_FORCE_NVFP4_MARLIN` | Force using NVFP4 Marlin fallback kernels even on Blackwell GPUs with native FP4 support | `false` |
```
Comment (b8zhong) [inline docs/references/environment_variables.md]: Can we remove this toggle for the sake of complexity, I don't think there is really a real use ase for it

File: python/sglang/srt/model_loader/weight_utils.py [hunk 2]
```diff
@@ -187,6 +187,15 @@ def get_quant_config(
         if not isinstance(hf_quant_config, dict):
             hf_quant_config = hf_quant_config.to_dict()
         hf_quant_config["packed_modules_mapping"] = packed_modules_mapping
+        if model_config.quantization.startswith("modelopt"):
```
Comment (b8zhong) [inline python/sglang/srt/model_loader/weight_utils.py]: Is this intentional or unintentional?

**Review by b8zhong — CHANGES_REQUESTED** [review]

**Comment by Godmook** [comment]
@b8zhong @BBuf  I’ve refactored the Marlin FP4 fallback to ensure it’s as self contained as possible. All logic for ModelOpt and compressed-tensors is now isolated in dedicated files, with the only change in modelopt_quant.py being a small dispatch block in get_quant_method for non-Blackwell GPUs.

I also did accuracy tests and the results are consistent with the previous implementation.

**Comment by Godmook** [comment]
@b8zhong @BBuf Ping! :) No rush at all!

**Comment by huxiang088** [comment]
@BBuf @b8zhong @Godmook  Is there any latest PR progress

**Comment by Godmook** [comment]
@huxiang088 I'm waiting for maintainers' approval, and if the CI passes afterward, the relanding will be complete.

**Comment by wxsms** [comment]
Can't wait for this to get merge! Thank you!

**Comment by Godmook** [comment]
Ping! @b8zhong @BBuf

**Comment by fy1214** [comment]
I think this is the best pr for w4(nvfp4)a16 RL !

**Review by b8zhong — APPROVED** [review]

File: test/registered/quant/test_nvfp4_marlin_fallback.py [hunk 3]
```diff
@@ -0,0 +1,788 @@
+"""Tests for NVFP4 Marlin fallback on non-Blackwell GPUs (SM75+)."""
+
+import unittest
+
+import torch
+
+from sglang.test.ci.ci_register import register_cuda_ci
+from sglang.test.test_utils import CustomTestCase
+
+register_cuda_ci(est_time=480, suite="stage-b-test-1-gpu-large")
```
Comment (AgainstEntropy) [inline test/registered/quant/test_nvfp4_marlin_fallback.py]: hi @Godmook , can you check the [test doc](https://github.com/sgl-project/sglang/tree/main/test#ci-registration) and update the CI registration accordingly?

Comment (Godmook) [reply test/registered/quant/test_nvfp4_marlin_fallback.py]: Yes, I checked that issue last night. Now I'm reviewing the test doc and updating the CI registration with the new version. Thanks for letting me know!

**Comment by Godmook** [comment]
@BBuf @AgainstEntropy @b8zhong  Heads up . I saw #25655 re-add NVFP4 Marlin on main, since this is a reland of my earlier #19652 (merged, then reverted in #22047) and it works same thing I thinks... Should I close it? or not? I'm busy these days so I don't have time to check this PR...

**Comment by Godmook** [comment]
> @BBuf @AgainstEntropy @b8zhong  Heads up . I saw #25655 re-add NVFP4 Marlin on main, since this is a reland of my earlier #19652 (merged, then reverted in #22047) and it works same thing I thinks... Should I close it? or not? I'm busy these days so I don't have time to check this PR...

Ping! @BBuf @AgainstEntropy @b8zhong :)

**Comment by Godmook** [comment]
Closing this PR.  superseded by #25655.

Just leaving the trail here for anyone curious about the history later: #19652 (merged → reverted in #22047) → this reland (open) → #25655 (same feature, re-implemented, merged). All roads lead to the same place eventually. 🙂

Thanks for getting it in!
